core/src/main/scala/org/apache/spark/MapOutputTracker.scala

```diff
@@ -132,13 +133,46 @@ private[spark] abstract class MapOutputTracker(conf: SparkConf) extends Logging
    *         describing the shuffle blocks that are stored at that block manager.
    */
   def getMapSizesByExecutorId(shuffleId: Int, reduceId: Int)
-  : Seq[(BlockManagerId, Seq[(BlockId, Long)])] = {
+      : Seq[(BlockManagerId, Seq[(BlockId, Long)])] = {
     logDebug(s"Fetching outputs for shuffle $shuffleId, reduce $reduceId")
-    val startTime = System.currentTimeMillis
+    val statuses = getStatuses(shuffleId)
+    // Synchronize on the returned array because, on the driver, it gets mutated in place
+    statuses.synchronized {
+      return MapOutputTracker.convertMapStatuses(shuffleId, reduceId, statuses)
+    }
+  }
+
+  /**
+   * Return statistics about all of the outputs for a given shuffle.
+   *
+   * @param shuffleId ID of the shuffle
+   * @param numPartitions number of map output partitions
+   */
+  def getStatistics(shuffleId: Int, numPartitions: Int): MapOutputStatistics = {
```

Comment:
its a little strange that you need to pass `numPartitions` in here, since that is a property of the map output.  Maybe instead of just taking  `shuffleId`, it should take a `shuffleHandle`, so you can get the number of map output partitions w/ `shuffleHandle.dependency.partioner.numPartitions`?